File: 40.组合总和-ii.cs
/*
 * @lc app=leetcode.cn id=40 lang=csharp
 *
 * [40] 组合总和 II
 */

// @lc code=start
public partial class Solution
{
    public IList<IList<int>> CombinationSum2(int[] candidates, int target)
    {
        IList<IList<int>> results = new List<IList<int>>();
        List<int> store = new List<int>();
        Array.Sort(candidates);
        Dfs(candidates, target, store, results, 0);
        return results;
    }

    public void Dfs(int[] candidates, int target, List<int> store, IList<IList<int>> results, int index)
    {
        if (target == 0 && store.Count != 0)
        {
            results.Add(store.ToList());
            return;
        }
        if (index == candidates.Length)
        {
            return;
        }
        if (target > 0)
        {
            store.Add(candidates[index]);
            Dfs(candidates, target - candidates[index], store, results, index + 1);
            store.RemoveAt(store.Count - 1);

            do
            {
                index++;
            } while (index < candidates.Length && candidates[index] == candidates[index - 1]);

            Dfs(candidates, target, store, results, index);
        }
    }
}
// @lc code=end

File: src/40.组合总和-ii.rs

/*
 * @lc app=leetcode.cn id=40 lang=rust
 *
 * [40] 组合总和 II
 */

// @lc code=start
impl Solution {
    pub fn combination_sum2(mut candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut results = Vec::new();
        let mut store = Vec::new();
        candidates.sort();
        dfs(&candidates, &mut results, &mut store, 0, target);
        return results;
    }
}

pub fn dfs(
    candidates: &Vec<i32>,
    mut results: &mut Vec<Vec<i32>>,
    mut store: &mut Vec<i32>,
    mut index: usize,
    target: i32,
) {
    if target == 0 && store.len() != 0 {
        results.push(store.to_owned())
    }
    if index >= candidates.len() {
        return;
    }
    if target > 0 {
        store.push(candidates[index]);
        dfs(
            &candidates,
            &mut results,
            &mut store,
            index + 1,
            target - candidates[index],
        );
        store.remove(store.len() - 1);

        loop {
            index += 1;
            if index >= candidates.len() || candidates[index] != candidates[index - 1] {
                break;
            }
        }
        dfs(&candidates, &mut results, &mut store, index, target);
    }
}
// @lc code=end